### PR TITLE
Bug #3001 - Facts searchable by host id and name

### PR DESCRIPTION
**Pull request by dLobatog**
Before, an api call such as foreman/api/v2/host/3232/facts would simply
return no facts, as the `id` field was only searching for a host
name. 

This PR fixes that and now the intermediate `host_id` field
allows to search by name and id by giving the SQL conditions directly to
scoped_search.

### CODE REVIEW

**Comment by domcleal** [comment]
@isratrade could you review this please when you get a chance?

File: test/functional/api/v1/fact_values_controller_test.rb [hunk 1]
```diff
@@ -6,14 +6,23 @@ class Api::V1::FactValuesControllerTest < ActionController::TestCase
     get :index, { }
     assert_response :success
     fact_values = ActiveSupport::JSON.decode(@response.body)
-    assert !fact_values.empty?
+    assert_not_empty fact_values
```
Comment (isratrade) [inline test/functional/api/v1/fact_values_controller_test.rb]: nitpick, I usually go with minitest's refute notation
`refute_empty`

File: test/functional/api/v2/fact_values_controller_test.rb [hunk 2]
```diff
@@ -6,14 +6,23 @@ class Api::V2::FactValuesControllerTest < ActionController::TestCase
     get :index, { }
     assert_response :success
     fact_values = ActiveSupport::JSON.decode(@response.body)
-    assert !fact_values.empty?
+    assert_not_empty fact_values
```
Comment (isratrade) [inline test/functional/api/v2/fact_values_controller_test.rb]: same nitpick here, refute_empty

File: test/functional/api/v2/fact_values_controller_test.rb [hunk 3]
```diff
@@ -6,14 +6,23 @@ class Api::V2::FactValuesControllerTest < ActionController::TestCase
...
   end
 
   test "should get facts for given host only" do
-    get :index, {:host_id => hosts(:one).to_param }
+    get :index, {:host_id => hosts(:one).name }
```
Comment (isratrade) [inline test/functional/api/v2/fact_values_controller_test.rb:13]: why did you change from `.to_param` to `.name`.  Just curious, both should work fine.

File: test/functional/api/v1/fact_values_controller_test.rb [hunk 4]
```diff
@@ -6,14 +6,23 @@ class Api::V1::FactValuesControllerTest < ActionController::TestCase
...
   end
 
   test "should get facts for given host only" do
-    get :index, {:host_id => hosts(:one).to_param }
+    get :index, {:host_id => hosts(:one).name }
```
Comment (isratrade) [inline test/functional/api/v1/fact_values_controller_test.rb:13]: did you change to name, because of the new method `search_by_host`?

File: test/functional/api/v1/fact_values_controller_test.rb [hunk 5]
```diff
@@ -6,14 +6,23 @@ class Api::V1::FactValuesControllerTest < ActionController::TestCase
...
+    assert_response :success
+    fact_values   = ActiveSupport::JSON.decode(@response.body)
+    expected_hash = FactValue.build_facts_hash(FactValue.find_all_by_host_id(hosts(:one).id))
```
Comment (isratrade) [inline test/functional/api/v1/fact_values_controller_test.rb]: I think `.where` is preferred over `.find_all_by_` if I remember correctly.  It think it's more efficient.  not sure.

File: app/models/fact_value.rb [hunk 6]
```diff
@@ -10,8 +11,8 @@ class FactValue < ActiveRecord::Base
   scoped_search :on => :value, :in_key=> :fact_name, :on_key=> :name, :rename => :facts, :complete_value => true
   scoped_search :on => :value, :default_order => true
   scoped_search :in => :fact_name, :on => :name, :complete_value => true, :alias => "fact"
-  scoped_search :in => :host, :on => :name, :rename => :host, :complete_value => true
```
Comment (isratrade) [inline app/models/fact_value.rb:14]: did this not work previously?  when you change the search code, you affect all the API's that use this search definition.  Would it be better to rename it to "host.something"

**Comment by isratrade** [comment]
@eLobato, looks good.  I haven't run the code yet, just gave my initial comments based on reviewing the code.

**Comment by dLobatog** [comment]
Cool, I'll make the changes. I forgot to mention this is basically how Hammer searches for Host facts, hence it's broken until the PR gets merged.

File: test/functional/api/v1/fact_values_controller_test.rb (hunk 4 above)
Comment (dLobatog) [inline test/functional/api/v1/fact_values_controller_test.rb:13]: Oh it's just a way of emphasizing that it can search by `Host` name and `Host` id now. I thought `to_param` doesn't make it that clear.

File: app/models/fact_value.rb (hunk 6 above)
Comment (dLobatog) [inline app/models/fact_value.rb:14]: It only worked for name, the new definition is just below and works for name and id to make Hammer users happy.  API users shouldn't get different results, except for users who had a host named "1" and another host with an id of "1", in this case they'd get two hosts in the results.

**Comment by isratrade** [comment]
Btw, I added an issue http://projects.theforeman.org/issues/4386 even though it doesn't relate to (scoped_search) to implement 'friendly_id' gem to simply all our login the UI and API by finding by id, name, label, etc.

File: test/functional/api/v1/fact_values_controller_test.rb (hunk 4 above)
Comment (isratrade) [inline test/functional/api/v1/fact_values_controller_test.rb:13]: @eLobato , agreed.  I didn't realize to_param = name in Host::Base

**Comment by domcleal** [comment]
@eLobato please rebase

**Comment by dLobatog** [comment]
Rebased

**Comment by domcleal** [comment]
Merged in e44f5c1 for Foreman 1.5.0, thanks @eLobato.
